### DISCUSSION TITLE
AdamW optimizer gives better loss convergence but poor image generation

### DISCUSSION THREAD

**Discussion by kobiso** [lucidrains/DALLE-pytorch#139] [Show and tell]
In DALLE paper, they used AdamW optimizer, LR warmup, and LR decay. (related PR: https://github.com/lucidrains/DALLE-pytorch/pull/138)
![image](https://user-images.githubusercontent.com/12422441/112940922-c7d23300-9168-11eb-9708-607e622e5413.png)
![image](https://user-images.githubusercontent.com/12422441/112940935-cc96e700-9168-11eb-941c-7cea96610b8f.png)

And I found LR decay actually helpful in https://github.com/lucidrains/DALLE-pytorch/discussions/131, but AdamW optimizer gives very weird results.
As shown in the below results, AdamW optimizer gives very fast loss decrease in image cross-entropy (CE) loss, while text CE loss is the same with Adam optimizer.
However, the visualization of AdamW optimizer shows corrupted images, which is so weird.
The low image CE loss means the model is predicting correct image tokens, but the generated image is poor.
I couldn't figured it out yet, so wanted to discuss about this with you folks 😃 .
What do you think about think about this?

## AdamW experiment
- Models
  - Orange color: Adam optimizer
  - Gray: AdamW optimizer
 
### Training graph
![image](https://user-images.githubusercontent.com/12422441/112941179-20093500-9169-11eb-8a1a-830e5a5ea229.png)

### Visualization
![image](https://user-images.githubusercontent.com/12422441/112941351-69598480-9169-11eb-9ba2-7f4b9ee28f13.png)

**Comment by Ki6an** [comment]
awesome! would using [madgrad](https://github.com/facebookresearch/madgrad) yield faster convergence.
>A best-of-both-worlds optimizer with the generalization performance of SGD and at least as fast convergence as that of Adam, often faster

**Reply by kobiso** [reply]
Yeap! Since the reason of poor generation was weight decay, MADGRAD could be a good option.
They mentioned "You may need to use a lower weight decay than you are accustomed to. Often 0.".
I will try it out :)

**Comment by kobiso** [comment]
## Cause of poor generation

**The cause of poor generation is weight decay.**
Even though i added https://github.com/lucidrains/DALLE-pytorch/blob/4aade13e04fee9a7614c157897a800f60b9a1d48/train_dalle.py#L146 discarding weight decay for bias and LayerNorm weight, the poor generation happened.
When I just discarded all the weight decay, the generation became normal.
Most importantly, **AdamW seems better at loss convergence than Adam**, see the experiment below.

### Experiment
- **Models**
  - Orange: Adam
  - Red: AdamW without weight decay

![image](https://user-images.githubusercontent.com/12422441/113368659-4feb4f00-939a-11eb-9885-08e04da386b8.png)
![image](https://user-images.githubusercontent.com/12422441/113368663-57125d00-939a-11eb-8d69-9f93b860bf81.png)

**Reply by afiaka87** [reply]
Filed an issue with regard to the weight decay here https://github.com/lucidrains/DALLE-pytorch/issues/170 @kobiso

**Comment by kobiso** [comment]
@Ki6an 
I did some experiments on [MADGRAD](https://github.com/facebookresearch/madgrad) optimizer, but I couldn't find any advantage or performance improvement compared to Adam 😅

**Reply by lucidrains** [reply]
@kobiso what kind of learning rate have you tried training DALL-E on with MadGrad? another researcher recommends that the learning rate is at least 10x what you would normally use with Adam

**Reply by kobiso** [reply]
@lucidrains 
I tried MADGRAD with lr 0.00045 (best in Adam), 0.001 (larger), 0.0001 (smaller).
And lr 0.00045 was the best among them.
But still, I didn't see any better performance compared to Adam trained model.
I will see if 10x performs well :)

**Comment by alexander-shustanov** [comment]
Hello! Great results! Could you say, how many epochs the model was trained?

**Comment by shizhediao** [comment]
Hi are there any updates?
Could I use adamw to optimize the training process?
Thanks!
